### PR TITLE
End-to-end data testing via blocks

### LINKED ISSUE #841: Refactor unit tests
### Description

Following #840, we need to add unit tests to the now unit-sized processing functions in the WeatherDataService.

### LINKED ISSUE #842: Update end-to-end testing to capture backend code coverage
### Description

In order to make our tests more robust and reliable, our end-to-end tests should compute code coverage. Some prototyping around this work has been started in the `mgwalker/change-testing` branch. This will help us identify the test gaps that end-to-end testing should cover.

### Acceptance criteria

- [ ] End-to-end tests report code coverage
- [ ] The CI/CD code coverage reporter step properly integrates the coverage information from end-to-end tests

### LINKED ISSUE #843: Identify the testing gaps in end-to-end testing
### Description

After #842, we will disable unit tests and capture code coverage solely with end-to-end tests. We will use this data to identify test gaps and update or create new tests to close those gaps.

This issue includes all three steps:
- disable unit testing
- identify test gaps
- fill the gaps

### Acceptance criteria

- [ ] Code coverage indicates that our end-to-end tests cover as much of the code as reasonably possible

### LINKED ISSUE #844: Re-enable unit tests and >=90% code coverage requirements
### Description

After #841 and #843, we need to re-enable unit tests and the minimum code coverage requirements in CI/CD.

### PR DESCRIPTION
**Pull request by greg-does-weather**
### End-to-end tests via blocks

This PR introduces the concept of treating our blocks as essentially API endpoints. Instead of trying to unit test everything in the back end with extensive, complicated, and fragile mocks, we can instead build the blocks in a functional Drupal instance using our API proxy to control the input data and verify what the block gives us back.

#### end-to-end block test base
- sets up autoloading based on Drupal settings
- loads the database connection from Drupal settings file and creates a connection
- creates a DataLayer
  - real HTTP client and database connection
  - mocked cache, route match interface
- creates a WeatherDataService
  - real DataLayer
  - mocked translation service, New Relic
- handles mocking the logger trait on blocks under test
- sends a request to the API proxy to enable the e2e bundle
- provides a helper method for setting the lat/lon of the location to load data from

##### Alerts
- Data structure is as expected
- Alerts covering the same forecast zone as the location are included
- Alerts covering the same fire zone as the location are included
- Alerts covering the same county zone as the location are included
- Alerts covering the same county SAME code as the location are included
- Alerts with geometry that intersects the geometry of the WFO grid of the location are included
- Does not include marine alerts
- Alerts are sorted correctly

##### Daily forecast
- Data structure is as expected
- The hours associated with the first day begin at the top of the next hour
- The hours associated with the first day end at 6am the next day
- All other days' hours begin at 6am and end at 6pm

##### Hourly forecast
- Data structure is as expected
- Hours are in the timezone of the location being queried
- Past hours are removed
- Past alerts are removed
- Data structure for embedded alerts is as expected
- Hourly alerts are properly aligned to their hour
- Hourly alerts are given the correct hourly duration

##### Hourly precipitation
- Past hours are removed
- Accumulations are rounded to the 1/100th of an inch

##### Observations (current conditions)
- Data structure is as expected
- `null` wind speed is returned as `null`
- `0` wind speed is returned as `0`
- wind speed is converted from km/h to mph
- if no heat index or wind chill, current temperature is reported as "feels like"
- if heat index is present, heat index is reported as "feels like"
- if wind chill is present and heat index is missing, wind chill is reported as "feels like"
- first observation station is used if observation is valid
- second observation station is used if first is invalid
- an error is returned if the first three observations are invalid

#### Alert trait
- moves `getEndTimeForAlert` to AlertUtility class
- create `getDurationText` method in AlertUtility class
  - updates AlertTrait to use this method instead of having it embedded

#### Daily forecast trait
- fixes a bug where past day periods could be returned (oops)

#### Data Layer
- adds a `preload` option to the constructor, defaults to `true`
  This is to support end-to-end testing. It prevents us from making a bunch of HTTP calls we don't need for a given test.
- guarantees that state abbreviation is upper-cased
- fixes a PHP warning when checking whether a property exists
  If you try to access a property that doesn't exist, that's a PHP warning. It still evaluates to falsey, but PHP warnings cause phpunit exceptions, so...
- reorders the concurrent API calls into alphabetical, just so they're in some kind of order

#### Hourly forecast trait
- adds `startRaw` and `endRaw` properties to hourly periods
  I don't remember why. Maybe we can revert this one.

#### Observation trait
- Fixes bugs revealed in end-to-end testing

- default URL to an empty string instead of null
- adds timestamp as an optional argument for DI

#### unit test block base
- handles implementing a mock logger trait
- creates a mock of the block under test to override the base `getLocation` method instead of relying on it to parse location information
- updates the `onLocationRoute` helper method to mock the `getLocation` method accordingly
- adds a `notOnLocationRoute` helper to mock the `getLocation` method accordingly
- adds a test method for if the WeatherDataService throws an exception – blocks get this test for free!

#### API-driven blocks
- verify they populate properly on point routes
- verify they populate properly on grid routes (legacy, but also necessary for configurable blocks)
- verify they return null if not on a point or grid route
- other tests as appropriate per block

#### Content-driven blocks
- override the base exception test, since they don't need it; they just assert true

#### Data service unit tests
- deletes most of the existing ones
- fully tests the DataLayer
- tests the utility classes (alerts, unit conversion, NewRelic)

### CI/CD

- shuffles the PHP unit testing task to after the image build, since our unit tests now rely on a fully-functioning Drupal
- similarly shuffles the test coverage task
- re-enables the unit testing and coverage tasks

### Misc
- deletes test data previously used by unit tests in favor of proxy data
- adds a end-to-end proxy bundle
- expands the API proxy UI to provide more information about the points available in a bundle
- fixes a bug in how `date:today` offsets were calculated
- fixes a nondeterministic unit test related to icons

---

closes #841
closes #842
closes #843 
closes #844

### CODE REVIEW

**Review by eric-gade — APPROVED** [review]
Divine and necessary work! I think this is an excellent design that will make work much easier going forward!
  
stom00 jpg!Large


## Code Review Checklist

This is an automated comment on every pull request requiring a review. A checked item is either an assertion that I have tested this item or an indication that I have verified it does not apply to this pull request.

## The Basics
- [x] Checks are passing
- [x] I read the code
- [x] I ran the code
- [ ] (if applicable) Post deploy steps are run
- [ ] (if applicable) I validated the change on the deployed version in <environment>

## Documentation
- [x] changes to “how we do things” are documented in READMEs
- [x] all new functions and methods are commented using plain language
- [x] any new modules added documented in modules.md

## Reliability
- [x] error handling exists for unusual or missing values
- [x] interactions with external systems are wrapped in try/except
- [x] functionality is tested with unit or integration tests
- [x] dependency updates in composer.json also got changed in composer-lock.json

## Infrastructure
- [x] all changes are auditable and documented via a script
- [x] it is clear who can and should run the script
- [ ] (if applicable) diagrams have been updated or added in PlantUML

## Accessibility
- [ ] New pages have been added to cypress-axe file so that they will be tested with our automated accessibility testing
- [ ] Meets WCAG 2.0 AA or 2.1 AA for Section 508 compliance
    - [ ] Site is keyboard accessible. All interactions can be accessed with a keyboard
    - [ ] Site is free of keyboard traps. The keyboard focus is never trapped in a loop
    - [ ] All form inputs have explicit labels
    - [ ] Form instructions are associated with inputs
    - [ ] All relevant images use an img tag
    - [ ] All images have appropriate alt attributes
    - [ ] Multimedia is tagged. All multimedia has appropriate captioning and audio description
    - [ ] Text has sufficient color contrast. All text has a contrast ratio of 4.5:1 with the background
    - [ ] Site never loses focus. Focus is always visible when moving through the page with the keyboard
    - [ ] Tab order is logical
    - [ ] Tables are coded properly. Tables have proper headers and column attributes
    - [ ] Headings are nested properly. Heading elements are nested in a logical way
    - [ ] Language is set. The language for the page is set
    - [ ] CSS is not required to use the page. The page makes sense with or without CSS
    - [ ] Links are unique and contextual. All links can be understood taken alone, e.g., ‘Read more - about 508’
    - [ ] Page titles are descriptive

## Device Matrix
- [x] firefox/gecko (renders correctly and user interactions work)
- [x] chrome/chromium/edge (renders correctly and user interactions work)
- [x] safari/webkit (renders correctly and user interactions work)
- [x] web page is readable and usable 
    - [ ] at 480px     (mobile)
    - [ ] at 640px     (tablet)
    - [ ] at 1024px    (desktop)

File: web/modules/weather_blocks/src/Plugin/Block/Test/CurrentConditionsBlock.php.test [hunk 1]
```diff
@@ -12,51 +12,122 @@ final class CurrentConditionsBlockTest extends Base
     /**
      * Common setup for all component tests.
      *
-     * Creates a mock of a weather data service and injects it into a new current
-     * conditions block object.
+     * Creates a mock of a weather data service and injects it into a new local
+     * alert block object.
```
[the remark this answers is not in the document]
Comment (greg-does-weather) [reply web/modules/weather_blocks/src/Plugin/Block/Test/CurrentConditionsBlock.php.test:16]: Oopsidoodle! You're right. Will push a fix in a moment.

File: web/modules/weather_blocks/src/Plugin/Block/Test/EndToEnd/00_EndToEndBase.php.test [hunk 2]
```diff
@@ -0,0 +1,153 @@
+<?php
```
[the remark this answers is not in the document]
Comment (greg-does-weather) [reply web/modules/weather_blocks/src/Plugin/Block/Test/EndToEnd/00_EndToEndBase.php.test:1]: Yeah, exactly. It's annoying! I think your approach was to remove the `.test` extension? That'd work too. Do you have a preference, because I don't. 😛

File: web/modules/weather_blocks/src/Plugin/Block/Test/EndToEnd/00_EndToEndBase.php.test [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+        $dataLayer = new DataLayer(
+            $httpClient,
+            $cache,
+            $database,
+            $this->routeMock,
+            false,
+        );
+
+        $nrPromise = new Promise();
```
[the remark this answers is not in the document]
Comment (greg-does-weather) [reply web/modules/weather_blocks/src/Plugin/Block/Test/EndToEnd/00_EndToEndBase.php.test:106]: Ooooh. I did not know that. I wondered at one point because such a thing ought to exist (maybe it's even in the A+ spec?), but then I just did the thing I knew how to do. I'm going to change this too just to be cleaner.

File: web/modules/weather_blocks/src/Plugin/Block/Test/EndToEnd/00_EndToEndBase.php.test [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+     *
+     * This is injected into the current conditions block object being tested with
+     * dependency injection.
+     */
+    protected $weatherData;
+
+    /**
+     * Common setup for all component tests.
+     */
+    protected function setUp($type = null): void
```
[the remark this answers is not in the document]
Comment (greg-does-weather) [reply web/modules/weather_blocks/src/Plugin/Block/Test/EndToEnd/00_EndToEndBase.php.test:57]: I think if it's not set by a child test, it'll throw a type error anyway. But we can be more explicit. Perhaps an issue for a debt ticket?

File: web/modules/weather_blocks/src/Plugin/Block/Test/EndToEnd/Observations/InvalidObservations.php.test [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+     */
+    public function testFirstObservationIsInvalid(): void
+    {
+        $this->onLocationRoute(35.198, -111.651);
+
+        $data = $this->block->build();
+
+        $this->assertSame(64, $data["temperature"]);
+    }
+
```
[the remark this answers is not in the document]
Comment (greg-does-weather) [reply web/modules/weather_blocks/src/Plugin/Block/Test/EndToEnd/Observations/InvalidObservations.php.test:45]: I thought about it, but I wasn't sure it gained us much confidence in the code. My gut feeling is that the only knew attestation would be *how* it iterates through observation stations, not that it actually does so since that's already attested?
